### PR TITLE
[ingress-nginx] Distroless kruise

### PR DESCRIPTION
**Pull request by mvasl**
## Description

kruise controller image is now based on distroless

## Why do we need it, and what problem does it solve?

We want our images to be distroless-based.

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: ingress-nginx
type: feature
summary: Kruise controller image is now based on a distroless image.
impact: kruise pod will be restarted.
impact_level: default
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've started the workflow for label 'e2e/run/static'...

File: modules/402-ingress-nginx/images/kruise-state-metrics/Dockerfile [hunk 1]
```diff
@@ -1,17 +1,32 @@
...
+RUN apk add --no-cache bash expat && rm -rf /var/cache/apk/*
 
-FROM $BASE_ALPINE
-RUN apk add --no-cache ca-certificates bash expat && \
-    rm -rf /var/cache/apk/*
+FROM $BASE_DISTROLESS
+COPY --from=deps /bin/bash /bin/bash
+COPY --from=deps /etc/bash/bashrc /etc/bash/bashrc
+COPY --from=deps /usr/lib/bash /usr/lib/bash/
+COPY --from=deps /usr/bin/xmlwf /usr/bin/xmlwf
```
Comment (RomanenkoDenys) [inline modules/402-ingress-nginx/images/kruise-state-metrics/Dockerfile:29]: Why we need it ?

File: modules/402-ingress-nginx/images/kruise/Dockerfile [hunk 2]
```diff
@@ -1,24 +1,38 @@
...
+RUN apk add --no-cache bash expat && rm -rf /var/cache/apk/*
 
-FROM $BASE_ALPINE
-RUN apk add --no-cache ca-certificates bash expat && \
-    rm -rf /var/cache/apk/*
+FROM $BASE_DISTROLESS
+COPY --from=deps /bin/bash /bin/bash
+COPY --from=deps /etc/bash/bashrc /etc/bash/bashrc
+COPY --from=deps /usr/lib/bash /usr/lib/bash/
+COPY --from=deps /usr/bin/xmlwf /usr/bin/xmlwf
```
Comment (RomanenkoDenys) [inline modules/402-ingress-nginx/images/kruise/Dockerfile:35]: Why we need it ?

**Review by aleksey-su — DISMISSED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've started the workflow for label 'e2e/run/openstack'...

**Comment by mvasl** [comment]
/e2e/abort openstack;Standard;containerd;1.25 6453269738 6453269738-1-con-1-25 /sys/deckhouse-oss/install:pr5993

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by mvasl** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.25 6453781334 6453781334-1-con-1-25 /sys/deckhouse-oss/install:pr5993 cloud-user@51.250.80.66

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by mvasl** [comment]
/e2e/abort yandex-cloud;WithoutNAT;containerd;1.25 6454548612 6454548612-1-con-1-25 /sys/deckhouse-oss/install:pr5993 cloud-user@158.160.96.120

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've started the workflow for label 'e2e/run/aws'...

**Comment by mvasl** [comment]
/e2e/abort aws;WithoutNAT;containerd;1.25 6455305715 6455305715-1-con-1-25 /sys/deckhouse-oss/install:pr5993 centos@3.76.16.153

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've started the workflow for label 'e2e/run/aws'...

File: modules/402-ingress-nginx/images/kruise/Dockerfile [hunk 3]
```diff
@@ -1,24 +1,30 @@
...
+    CGO_ENABLED=0 \
+    GOOS=linux \
+    GOARCH=amd64
 
 COPY patches/disable-controllers.patch /
 COPY patches/disable-jobs.patch /
 COPY patches/stick-to-maxunavailable.patch /
 
-RUN git clone --depth 1 --branch v${KRUISE_CONTROLLER_VERSION} https://github.com/openkruise/kruise.git && \
+RUN git -c advice.detachedHead=false clone --depth 1 --branch v${KRUISE_CONTROLLER_VERSION} ${SOURCE_REPO}/openkruise/kruise.git && \
```
Comment (RomanenkoDenys) [inline modules/402-ingress-nginx/images/kruise/Dockerfile:19]: why do we need this option here ?

Comment (RomanenkoDenys) [reply modules/402-ingress-nginx/images/kruise/Dockerfile:19]: -c advice.detachedHead=false ?

**Review by RomanenkoDenys — COMMENTED** [review]
some questions

File: modules/402-ingress-nginx/images/kruise/Dockerfile (hunk 3 above)
Comment (mvasl) [reply modules/402-ingress-nginx/images/kruise/Dockerfile:19]: git exits with non-zero code without it if you hit detached state after clone without this iirc

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've started the workflow for label 'e2e/run/openstack'...

File: modules/402-ingress-nginx/images/kruise-state-metrics/werf.inc.yaml [hunk 4]
```diff
@@ -0,0 +1,27 @@
+---
```
Comment (RomanenkoDenys) [inline modules/402-ingress-nginx/images/kruise-state-metrics/werf.inc.yaml:1]:
```suggestion
{{- $branch := "master }}
{{- $commit := "b081f2ae9e011fd92fd23f6efa209601f5a20a01" }}

---
```

File: modules/402-ingress-nginx/images/kruise-state-metrics/werf.inc.yaml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+artifact: kruise-state-metrics
+from: {{ $.Images.BASE_GOLANG_19_BULLSEYE }}
+shell:
+  setup:
+    - export "KRUISE_STATE_METRICS_BRANCH=master"
+    - export "KRUISE_STATE_METRICS_COMMIT=b081f2ae9e011fd92fd23f6efa209601f5a20a01"
+    - export "CGO_ENABLED=0"
+    - export "GOOS=linux"
+    - export "GOARCH=amd64"
+    - git clone --depth 1 --branch ${KRUISE_STATE_METRICS_BRANCH} {{ $.SOURCE_REPO }}/openkruise/kruise-state-metrics.git
```
Comment (RomanenkoDenys) [inline modules/402-ingress-nginx/images/kruise-state-metrics/werf.inc.yaml:11]:
```suggestion
    - git clone --depth 1 --branch {{ $branch }} {{ $.SOURCE_REPO }}/openkruise/kruise-state-metrics.git
```

File: modules/402-ingress-nginx/images/kruise-state-metrics/werf.inc.yaml [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    - cd kruise-state-metrics
+    - git checkout ${KRUISE_STATE_METRICS_COMMIT}
```
Comment (RomanenkoDenys) [inline modules/402-ingress-nginx/images/kruise-state-metrics/werf.inc.yaml:13]:
```suggestion
    - git checkout {{ $commit }}
```

File: modules/402-ingress-nginx/images/kruise-state-metrics/werf.inc.yaml [hunk 7]
```diff
@@ -0,0 +1,27 @@
+---
+artifact: kruise-state-metrics
+from: {{ $.Images.BASE_GOLANG_19_BULLSEYE }}
+shell:
+  setup:
+    - export "KRUISE_STATE_METRICS_BRANCH=master"
+    - export "KRUISE_STATE_METRICS_COMMIT=b081f2ae9e011fd92fd23f6efa209601f5a20a01"
```
Comment (RomanenkoDenys) [inline modules/402-ingress-nginx/images/kruise-state-metrics/werf.inc.yaml:7]:
```suggestion
```

File: modules/402-ingress-nginx/images/kruise/werf.inc.yaml [hunk 8]
```diff
@@ -0,0 +1,34 @@
+---
```
Comment (RomanenkoDenys) [inline modules/402-ingress-nginx/images/kruise/werf.inc.yaml:1]:
```suggestion
{{- $version := "1.4.0" }}

---
```

File: modules/402-ingress-nginx/images/kruise/werf.inc.yaml [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        - '**/*'
+    includePaths:
+      - patches
+shell:
+  setup:
+    - export "KRUISE_CONTROLLER_VERSION=1.4.0"
+    - export "CGO_ENABLED=0"
+    - export "GOOS=linux"
+    - export "GOARCH=amd64"
+    - git clone --depth 1 --branch v${KRUISE_CONTROLLER_VERSION} {{ $.SOURCE_REPO }}/openkruise/kruise.git
```
Comment (RomanenkoDenys) [inline modules/402-ingress-nginx/images/kruise/werf.inc.yaml:18]:
```suggestion
    - export "CGO_ENABLED=0"
    - export "GOOS=linux"
    - export "GOARCH=amd64"
    - git clone --depth 1 --branch v{{ $version }} {{ $.SOURCE_REPO }}/openkruise/kruise.git
```

File: modules/402-ingress-nginx/images/kruise/werf.inc.yaml [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    - cd kruise
+    - git apply /patches/*.patch
```
Comment (RomanenkoDenys) [inline modules/402-ingress-nginx/images/kruise/werf.inc.yaml:20]:
```suggestion
    - find /patches -name '*.patch' -exec git apply {} \;
```

**Review by RomanenkoDenys — CHANGES_REQUESTED** [review]
some nits

File: modules/402-ingress-nginx/images/kruise/werf.inc.yaml (hunk 10 above)
Comment (mvasl) [reply modules/402-ingress-nginx/images/kruise/werf.inc.yaml:20]: Why?

**Review by RomanenkoDenys — APPROVED** [review]
LGTM

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @mvasl. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by mvasl** [comment]
Manual test: looks ok to me, controller works and metrics are there as well
Снимок экрана 2023-10-13 в 12 30 34
Снимок экрана 2023-10-13 в 12 29 53

**Review by name212 — APPROVED** [review]

**Pull request merged**